src/metabase/lib/util/match/impl.cljc

```diff
@@ -26,15 +26,23 @@
   {:pre [(fn? match-fn) (vector? clause-parents)]}
   (cond
     (map? form)
-    (mapcat (fn [[k v]]
-              (match-fn (conj clause-parents k) v))
-            form)
+    (reduce-kv (fn [acc k v]
+                 (if-let [match (match-fn (conj clause-parents k) v)]
+                   ;; Deliberately not using into to avoid converting to transient and back.
```

Comment:
What about doing the entire `reduce-kv` over a transient `acc` then?

Reply:
I've tried that and it was less efficient. Apparently, the match-fn matches so rarely that the empty vector is never appended to.